### DISCUSSION TITLE
认为应该添加anime4k的A+A方案

### DISCUSSION THREAD

**Discussion by kbdrla** [dyphire/mpv-config#97] [Ideas]
之前试用了懒人包，不是太喜欢，又自己搞了anime专用版，没有ui又觉得鸡肋，直到这几天用了作者的版本，很是喜欢。
感觉最好的还是最近播放和文件夹浏览功能，这里提几个对优化的建议。
通常来说mpv的用户大部分有anime播放的需求，我之前也是因为有这方面的想法才知道mpv这个软件的。而本播放器的通用性强，针对性太弱了，或者说需要使用者进行更进一步的设置方面优化定制。
这里我认为有几点设置是不合理的：
1、默认按键1-8的设置应该修改为alt+数字。理由：几乎没什么人会去调整这些参数，即便会调整，频率也会相当低，可以说90%的用户只会调整一次后就不再使用这方面的功能。
2、修改slang参数。理由：大部分人是播放anime和外国电影居多，当然也有人看国产的多，即便是国产，也基本只有单国语音轨。这里优先度建议设置为ja,jap,en,eng,chs,sc,zh-CN,zh-Hans,cht,tc,zh-Hant,zh-HK,zh-TW,chi,zho,zh
3、HDR峰值设置修改为40 60 80 100或者针对不同hdr给出方案（设备有限，我不想研究）。理由：我的电脑是联想拯救者r9000p，所以屏幕是HDR400 分辨率2k。正常来说低于hdr1000的都是假的hdr，但想必大多数人没有好的hdr屏幕，至少我是没有的。在hdr400的情况下，配置文件设置hdr100显得过于暗淡，这个数值我测试过越低越亮，但也不是越低越好，太低会很假，70-80算是一个合适的数值。
4、常规配置组的进一步设置优化，针对1080p分辨率以上和以下的动画以及电影，针对2160p的动画和电影，针对HDR。理由：这几种几乎包含了用户能够播放的视频类型。anime4k给出了低中高三种优化方案，我下面写的是高端配置使用的。
代码我只是抄写修改了anime4k官方建议的版本，一个是A+A一个是B+B，分别适用于常规1080p分辨率和老番，配置可以就无脑一直用A+A方案就行。至于电影我研究的不多，所以就用默认启动配置。
#针对1080p动画，可以将色彩和清晰度极大提升，不亚于看2160p。适合配置显卡3070以上。官方A+A方案
[Anime4K]
 profile-desc=Anime4K
 profile-desc=适用于大多数动画
 glsl-shaders="~~/shaders/Anime4K/glsl/Restore/Anime4K_Clamp_Highlights.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_Restore_CNN_VL.glsl;~~/shaders/Anime4K/glsl/Upscale/Anime4K_Upscale_CNN_x2_VL.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_Restore_CNN_M.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_AutoDownscalePre_x2.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_AutoDownscalePre_x4.glsl;~~/shaders/Anime4K/glsl/Upscale/Anime4K_Upscale_CNN_x2_M.glsl"
#适用于1080p分辨率以下的老番，目前测试我觉得效果并不好，远不如无脑用上面的方案。而且我也不喜欢看老番，就算是那些经典老番也早就有人针对性优化出了更清晰的版本片源，比如龙珠tv版就有1080p的。同样测试过MPC播放器使用N卡的VRS，效果比上面的A+A方案差，还是无脑用A+A方案最好。
[Anime4K+]
 profile-desc=Anime4K+
 profile-desc=适用于 SD 动画
 glsl-shaders="~~/shaders/Anime4K/glsl/Restore/Anime4K_Clamp_Highlights.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_Restore_CNN_Soft_VL.glsl;~~/shaders/Anime4K/glsl/Upscale/Anime4K_Upscale_CNN_x2_VL.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_AutoDownscalePre_x2.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_AutoDownscalePre_x4.glsl;~~/shaders/Anime4K/glsl/Restore/Anime4K_Restore_CNN_Soft_M.glsl;~~/shaders/Anime4K/glsl/Upscale/Anime4K_Upscale_CNN_x2_M.glsl;~~/shaders/FSRCNNX_x2_16-0-4-1_anime_enhance.glsl"

因为设备有限，我测试了1080p的动画，效果是很好的。虽然文件夹里有auto，但作者没开auto这两个着色器，导致画面效果差很多。这两个着色器作用是将画面放大后再次缩小，从而提升画面清晰度（好像是这样）。2160p的动画，有效果，但是不强，属于锦上添花。分辨率的提升比软件优化效果要强的多，最先追求的应该就是分辨率，然后才是优化。电影方面我没太多研究，希望有人专门测试吧，我对anime需求更大，电影要下载的话2160p的资源也更多，所以我认为除非喜欢看古早片或者电影噪点太多，这方面不需要太过折腾。
快捷键可以设置动画1080p和老番为：1和ctrl+1，电影：2和ctrl+2，通常模式3，2160p动画4，2160p电影5，HDR电影6，全部清空着色器0或者del。在配置文件当中可以写上中等配置和高等配置两种优化方案，至于低等配置我认为不适合折腾。如果有4k屏幕，并且喜欢折腾的话，2160p的动画也可以尝试调试出优化方案，只是我的设备限制，这方面看不出太大区别。
关于插帧：我电脑配置不太行，笔记本开了A+A的配置就差不多了，能不掉帧就好，插帧开了很卡，我也没这方面需求。高端用户可以研究，目前最常见的就是svp还有一个rf。
madvr对比：madvr太折腾了，而且我感觉效果虽然有时候挺强的，但对于不是投屏和设备不是太强的用户，这方面没必要追求，配置好的mpv播放器效果足够使用了。

**Comment by kbdrla** [comment]
更进一步我认为只要按两个键就可以完美使用，一个是动画模式，一个是电影模式，可以设置成1和2。mpv应该是没法方便出动画和电影的区别，毕竟不管哪个都是同样的格式。所以这点只能认为区别，但是mpv有甄别视频分辨率，帧率和HDR的功能。根据这个功能就可以配置出打开一个视频，就针对性使用该视频分辨率的方案。不过这里有个问题就是设备的配置问题，可能需要设置出低中高三种方案。
至于分辨率甄别，代码我抄的

[HDR_or_21:9]
profile-desc=cond:(p["video-params/primaries"]=="bt.2020" or p["video-params/aspect"]>=2.0)
blend-subtitles=no
# 使字幕输出在黑边上
sub-ass-force-margins=yes

[SDR_and_16:9]
profile-desc=cond:(p["video-params/primaries"]~="bt.2020" and p["video-params/aspect"]<2.0)
blend-subtitles=yes
sub-ass-force-margins=no

# 以下根据分辨率和视频帧率设置不同的升/降阶算法，以期获得在能不掉帧的情况下达到最好的效果

# 创建一个基于gpu-hq，用于将一些参数回归到默认的profile：
[high_quality]
profile=gpu-hq
## !!! 使用  fbo-format=rgba16hf  可能能提高效率
fbo-format=rgba32f
## !!! 使用  fbo-format=rgba16hf  可能能提高效率
vd-lavc-threads=0

# 1080 * 1.414213 / 4 = 381.8
[low] # 240p 360p，启用两次倍增以达到4倍.
profile-desc=cond:(p["video-params/w"]<=678 and p["video-params/h"]<=381)
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/KrigBilateral.glsl"
glsl-shaders-append="~~/shaders/nnedi3-nns32-win8x4.hook"
glsl-shaders-append="~~/shaders/nnedi3-nns32-win8x4.hook"

# 1080 / 16 * 9 = 607.5
[SD] # 480p 576p
profile-desc=cond:((p["video-params/w"]<1080 and p["video-params/h"]<608) and (p["video-params/w"]>678 or p["video-params/h"]>381))
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/KrigBilateral.glsl"
glsl-shaders-append="~~/shaders/nnedi3-nns64-win8x4.hook"

# 1080 * 1.414213 / 2 = 763.7
[HD30] # 640p 720p 30fps
profile-desc=cond:((p["video-params/w"]<1358 and p["video-params/h"]<764) and (p["video-params/w"]>=1080 or p["video-params/h"]>=608) and p["estimated-vf-fps"]<31)
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/KrigBilateral.glsl"
glsl-shaders-append="~~/shaders/nnedi3-nns32-win8x4.hook"

[HD60] # 640p 720p 60fps
profile-desc=cond:((p["video-params/w"]<1358 and p["video-params/h"]<764) and (p["video-params/w"]>=1080 or p["video-params/h"]>=608) and p["estimated-vf-fps"]>=31)
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/KrigBilateral.glsl"
glsl-shaders-append="~~/shaders/SSimSuperRes.glsl"
# SSimSuperRes推荐关闭sigmoid-upscaling。所以在其他不用SSSR的profile中都要重新设为yes（通过重新设定一遍profile=gpu-hq）
sigmoid-upscaling=no

[KrigBilateral] # 另一个profile的例子
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/KrigBilateral.glsl"

[FHD] # 1080p
profile-desc=cond:((p["video-params/w"]<=1920 and p["video-params/h"]<=1080) and (p["video-params/w"]>=1358 or p["video-params/h"]>=764))
profile=KrigBilateral

[2K30] # 1440p
profile-desc=cond:((p["video-params/w"]<=2560 and p["video-params/h"]<=1440) and (p["video-params/w"]>1920 or p["video-params/h"]>1080) and p["estimated-vf-fps"]<31)
profile=KrigBilateral

# 4K videos
[4K30] # low fps: use SSIM to downscale
profile-desc=cond:((p["video-params/w"]>2560 or p["video-params/h"]>1440) and p["estimated-vf-fps"]<31)
profile=high_quality
glsl-shaders-clr
glsl-shaders-append="~~/shaders/SSimDownscaler.glsl"
# SSimDownscaler要求。需要在其他profile中重新设为yes（通过重新设定一遍profile=gpu-hq）
linear-downscaling=no

[UHD60] # high fps: use no glsl-shaders
profile-desc=cond:((p["video-params/w"]>1920 or p["video-params/h"]>1080) and p["estimated-vf-fps"]>=31)
glsl-shaders-clr
# 如果是1080p屏幕播放4K，可以使用https://gist.github.com/bjin/15f307e7a1bdb55842bbb663ee1950ed
# glsl-shaders-append="~~/shaders/acme-0.5x.hook"
# 可以绕过chroma upscaling，并且把之后颜色转换的计算量从4K级降到1080p级。
fbo-format=auto
vd-lavc-threads=16
# 4K HEVC HDR 60fps软解压力很大，默认vd-lavc-threads=0自动识别逻辑核数，分配核数+1个线程，但往往不够，不能将CPU性能压榨到极限，强行分配更多线程数以提高软解性能。
# 或者直接开硬解，如果显卡支持的话
hwdec=yes

**Comment by dyphire** [comment]
不感兴趣，anime4k 的 CNN 放大算法劣化效果严重。默认配置中不会加入会引入劣化过多的预设，shaders 文件夹中 anime4k 内容是全的，任何人都可以根据个人喜好修改组合

基于同样的理由该配置永远不会引入补帧相关，但感兴趣的同样可以自行集成

PS：你建议的组合和自动配置有很多槽点，在这儿就不一一指出了（
